src/openms/source/ANALYSIS/TARGETED/MetaboTargetedTargetDecoy.cpp

```diff
@@ -0,0 +1,273 @@
+    std::vector<MetaboTargetedTargetDecoy::MetaboTargetDecoyMassMapping> mappings;
+    for (const auto& it : identifier)
+    {
+      MetaboTargetedTargetDecoy::MetaboTargetDecoyMassMapping mapping;
+      mapping.identifier = it;
+      auto it_target = rmts.begin();
+      while ((it_target = find_if(it_target,
+                                  rmts.end(),
+                                  [&it](ReactionMonitoringTransition &rts)
+                                  {
+                                    return rts.getMetaValue("m_ids_id") == it &&
+                                           rts.getDecoyTransitionType() == ReactionMonitoringTransition::TARGET;
+                                  })) != rmts.end())
+      {
+        mapping.target_product_masses.emplace_back(it_target->getProductMZ());
+        mapping.target_compound_ref = it_target->getCompoundRef();
+        ++it_target;
+      }
+      auto it_decoy = rmts.begin();
+      while ((it_decoy = find_if(it_decoy,
+                                 rmts.end(),
+                                 [&it](ReactionMonitoringTransition &rts)
+                                 {
+                                   return rts.getMetaValue("m_ids_id") == it &&
+                                          rts.getDecoyTransitionType() == ReactionMonitoringTransition::DECOY;
+                                 })) != rmts.end())
+      {
+        mapping.decoy_product_masses.emplace_back(it_decoy->getProductMZ());
+        mapping.decoy_compound_ref = it_decoy->getCompoundRef();
+        ++it_decoy;
+      }
+      mappings.emplace_back(mapping);
+    }
+    return mappings;
+  }
+
+  void MetaboTargetedTargetDecoy::resolveOverlappingTargetDecoyMassesByIndividualMassShift(TargetedExperiment& t_exp, std::vector<MetaboTargetedTargetDecoy::MetaboTargetDecoyMassMapping>& mappings, const double& mass_to_add)
+  {
+    // compare targets and decoy masses
+    for (auto &it : mappings)
+    {
+      std::vector<double> intersection;
+      std::vector<double> replace_intersection;
+      std::vector<double> target_product_masses = it.target_product_masses;
+      std::vector<double> decoy_product_masses = it.decoy_product_masses;
+      std::sort(target_product_masses.begin(), target_product_masses.end());
+      std::sort(decoy_product_masses.begin(), decoy_product_masses.end());
+
+      std::set_intersection(target_product_masses.begin(),
+                            target_product_masses.end(),
+                            decoy_product_masses.begin(),
+                            decoy_product_masses.end(),
+                            std::back_inserter(intersection));
+
+      if (!intersection.empty())
+      {
+        std::transform(intersection.begin(),
+                       intersection.end(),
+                       std::back_inserter(replace_intersection),
+                       [mass_to_add](double d) -> double { return d + mass_to_add; });
```

Comment:
I tried my best :D 

But I think it is ok, even if it happens it should only happen on a limited set of decoys.
Even some kind of overlap is not too bad for scoring, since some of the metabolites might also overlap in some of the transitions.